### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [542]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

In [543]:
#Verified data is complete
purchase_data.count()

Purchase ID    780
SN             780
Age            780
Gender         780
Item ID        780
Item Name      780
Price          780
dtype: int64

In [544]:
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [998]:
playerCount = purchase_data["SN"].nunique()
playerCount_df = pd.DataFrame({
    "Total Players" : [playerCount]
})

#Formating
th_props = [
  ('font-size', '14px'),
  ('text-align', 'center'),
  ('font-weight', 'bold'),
  ('color', '#F7F5FB'),
  ('background-color', '#324A5F')
  ]

td_props = [
  ('font-size', '13px'),
    ('text-align', 'center')
  ]

styles = [
  dict(selector="th", props=th_props),
  dict(selector="td", props=td_props)
  ]


(playerCount_df.style
    .set_table_styles(styles))

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [999]:
#Introduciong data into a df

purchasingAnalysis_df = pd.DataFrame({
    "Number of Unique Items" : purchase_data["Item Name"].nunique(),
    "Average Price" : [purchase_data["Price"].mean()],
    "Number of Purchases" : purchase_data["Purchase ID"].nunique(),
    "Total Revenue" : [purchase_data["Price"].sum()]
})


#Formating
th_props = [
  ('font-size', '14px'),
  ('text-align', 'center'),
  ('font-weight', 'bold'),
  ('color', '#F7F5FB'),
  ('background-color', '#324A5F')
  ]

td_props = [
  ('font-size', '13px'),
    ('text-align', 'center')
  ]

styles = [
  dict(selector="th", props=th_props),
  dict(selector="td", props=td_props)
  ]


(purchasingAnalysis_df.style
    .format({'Average Price': "${:.2f}"})
     .format({'Total Revenue': "${:.2f}"})
    .set_table_styles(styles))

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [1000]:
#Eliminating duplicates
purchase_data_df = pd.DataFrame(purchase_data)
purchase_data_df_nd = purchase_data.drop_duplicates(subset = "SN")

#Creating DF
gender_dem_df = pd.DataFrame ({
    "Total Count" : purchase_data_df_nd["Gender"].value_counts(),
    "Percentage of Players" : purchase_data_df_nd["Gender"].value_counts() * 100 / playerCount

})

#Formating
th_props = [
  ('font-size', '14px'),
  ('text-align', 'center'),
  ('font-weight', 'bold'),
  ('color', '#F7F5FB'),
  ('background-color', '#324A5F')
  ]

td_props = [
  ('font-size', '13px'),
    ('text-align', 'center')
  ]

styles = [
  dict(selector="th", props=th_props),
  dict(selector="td", props=td_props)
  ]


(gender_dem_df.style
    .format({'Percentage of Players': "{:.2f}%"})
    .set_table_styles(styles))



## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [1020]:
#Filtering DF for male users
maleF = purchase_data_df["Gender"] == "Male"
purchase_data_male = purchase_data_df.loc[maleF]

countM = purchase_data_male["Purchase ID"].count()
meanM = purchase_data_male["Price"].mean()
sumM = purchase_data_male["Price"].sum()
avgpurM = sumM/gender_nd["Male"]

#Filtering DF for female users
femaleF = purchase_data_df["Gender"] == "Female"
purchase_data_female = purchase_data_df.loc[femaleF]

countF = purchase_data_female["Purchase ID"].count()
meanF = purchase_data_female["Price"].mean()
sumF = purchase_data_female["Price"].sum()
avgpurF = sumF/gender_nd["Female"]

#Filtering DF for other users
otherF = purchase_data_df["Gender"] == "Other / Non-Disclosed"
purchase_data_other = purchase_data_df.loc[otherF]

countO = purchase_data_other["Purchase ID"].count()
meanO = purchase_data_other["Price"].mean()
sumO = purchase_data_other["Price"].sum()
avgpurO = sumO/gender_nd["Other / Non-Disclosed"]


#Creating DF (Maintained MFO order for consistency)
purchasing_analysis_g_df = pd.DataFrame ({
    "Gender" : ["Male", "Female", "Other / Non-Disclosed"],
    "Purchase Count" : [countM, countF, countO],
    "Average Purchase Price" : [meanM, meanF, meanO],
    "Total Purchase Value" : [sumM, sumF, sumO],
    "Avg Total Purchase per Person" : [avgpurM, avgpurF, avgpurO]
    
})

purchasing_analysis_g_df = purchasing_analysis_g_df.set_index("Gender")

#Formating
th_props = [
  ('font-size', '14px'),
  ('text-align', 'center'),
  ('font-weight', 'bold'),
  ('color', '#F7F5FB'),
  ('background-color', '#324A5F')
  ]

td_props = [
  ('font-size', '13px'),
    ('text-align', 'center')
  ]

styles = [
  dict(selector="th", props=th_props),
  dict(selector="td", props=td_props)
  ]


(purchasing_analysis_g_df.style
    .format({'Average Purchase Price': "${:.2f}"})
    .format({'Total Purchase Value': "${:,.2f}"})
    .format({'Avg Total Purchase per Person': "${:,.2f}"})
    .set_table_styles(styles))


## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [1023]:
bins =[0, 9, 14, 19, 24, 29, 34, 39, 99]

labels = ["<10","10-14","15-19", "20-24", "25-29", "30-34", "35-39",
         "40+"]

purchase_data_df_nd["Age Ranges"] = pd.cut(
    purchase_data_df_nd["Age"],
    bins,
    labels=labels)

purchase_data_male_nd = purchase_data_df_nd.loc[maleF]
purchase_data_female_nd = purchase_data_df_nd.loc[femaleF]

summary_demo_df = pd.DataFrame({
    
    "Total Count" : purchase_data_df_nd.groupby(
        "Age Ranges")["Age"].count(),
    "Percentage of Players": purchase_data_df_nd.groupby(
        "Age Ranges")["Age"].count() / playerCount * 100,
    "Males" : purchase_data_male_nd.groupby(
        "Age Ranges")["Age"].count(),
    "Females" : purchase_data_female_nd.groupby(
        "Age Ranges")["Age"].count()
    
})
#Formating
th_props = [
  ('font-size', '14px'),
  ('text-align', 'center'),
  ('font-weight', 'bold'),
  ('color', '#F7F5FB'),
  ('background-color', '#324A5F')
  ]

td_props = [
  ('font-size', '13px'),
    ('text-align', 'center')
  ]

styles = [
  dict(selector="th", props=th_props),
  dict(selector="td", props=td_props)
  ]


(summary_demo_df.style
    .format({'Percentage of Players': "{:.2f}%"})
    .set_table_styles(styles))


/Users/fer/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [1006]:
bins =[0, 9, 14, 19, 24, 29, 34, 39, 99]

labels = ["<10","10-14","15-19", "20-24", "25-29", "30-34", "35-39",
         "40+"]

purchase_data_df["Age Ranges"] = pd.cut(
    purchase_data_df["Age"],
    bins,
    labels=labels)

summary_ages_df = pd.DataFrame({
    
    "Purchase Count": purchase_data_df.groupby(
    "Age Ranges")["Purchase ID"].count(),
    
    "Average Purchase Price": purchase_data_df.groupby(
    "Age Ranges")["Price"].mean(),
                                
    "Total Purchase Value" : purchase_data_df.groupby(
    "Age Ranges")["Price"].sum(),
    
    "Avg Total Purchase per Person" : summary_ages_df ["Total Purchase Value"] / age_summary ["Total Count"],
     
})

#Formating
th_props = [
  ('font-size', '14px'),
  ('text-align', 'center'),
  ('font-weight', 'bold'),
  ('color', '#F7F5FB'),
  ('background-color', '#324A5F')
  ]

td_props = [
  ('font-size', '13px'),
    ('text-align', 'center')
  ]

styles = [
  dict(selector="th", props=th_props),
  dict(selector="td", props=td_props)
  ]


(summary_ages_df.style
    .format({'Average Purchase Price': "${:.2f}"})
    .format({'Total Purchase Value': "${:,.2f}"})
    .format({'Avg Total Purchase per Person': "${:.2f}"})
    .set_table_styles(styles))




## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [880]:
# purchase_data_df["SN"].value_counts()

top_spenders_df = pd.DataFrame({
    "Purchase Count" : purchase_data_df.groupby("SN")["Purchase ID"].count(),
    "Average Purchase Price" : purchase_data_df.groupby("SN")["Price"].sum()  
    / purchase_data_df.groupby("SN")["Purchase ID"].count(),
    "Total Purchase Value" : purchase_data_df.groupby("SN")["Price"].sum(),
})


top_spenders_df = top_spenders_df.sort_values(ascending = False, by = "Total Purchase Value").head(5)


#Formating
th_props = [
  ('font-size', '14px'),
  ('text-align', 'center'),
  ('font-weight', 'bold'),
  ('color', '#F7F5FB'),
  ('background-color', '#324A5F')
  ]

td_props = [
  ('font-size', '13px'),
    ('text-align', 'center')
  ]

styles = [
  dict(selector="th", props=th_props),
  dict(selector="td", props=td_props)
  ]


(top_spenders_df.style
    .format({'Average Purchase Price': "${:.2f}"})
    .format({'Total Purchase Value': "${:.2f}"})
    .set_table_styles(styles))


## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame

  

In [997]:
top_items_df = pd.DataFrame({
    "Item Purchase Count" : purchase_data_df.groupby("Item ID"
                                          )["Item ID"].count(),
    "Total Purchase Value": purchase_data_df.groupby("Item ID")[
        "Price"].sum(),
    "Item Price" : purchase_data_df["Price"],
    "Item Name" : purchase_data_df["Item Name"],
    "Item ID" : purchase_data_df["Item ID"]
 })
top_items_df = top_items_df.sort_values(
    by = "Total Purchase Value", ascending = False)
top_items_df.head(93)

Item Purchase Count  Total Purchase Value  Item Price  \
92                  13.0                 59.99        3.94   
178                 12.0                 50.76        4.60   
82                   9.0                 44.10        4.40   
145                  9.0                 41.22        1.33   
103                  8.0                 34.80        1.33   
59                   8.0                 33.84        3.03   
108                  9.0                 31.77        1.75   
78                   7.0                 30.80        4.60   
72                   8.0                 30.16        4.23   
132                  9.0                 28.99        4.58   
60                   8.0                 28.32        2.38   
179                  6.0                 26.88        4.23   
160                  6.0                 26.40        2.38   
25                   6.0                 26.10        4.23   
37                   8.0                 25.28        2.05   
139                  5.0                 24.70        2.45   
128                  5.0                 24.55        4.14   
138                  5.0                 23.70        3.58   
87                   5.0                 22.15        3.40   
159                  7.0                 21.56        3.94   
136                  6.0                 21.48        3.03   
65                   5.0                 21.20        2.22   
102                  6.0                 20.64        4.76   
133                  5.0                 20.45        1.28   
39                   5.0                 19.70        3.14   
111                  4.0                 19.56        4.75   
165                  4.0                 19.44        1.09   
75                   8.0                 19.36        1.29   
52                   4.0                 19.36        3.54   
157                  4.0                 19.20        3.53   
..                   ...                   ...         ...   
140                  5.0                 14.70        3.93   
64                   3.0                 14.46        2.92   
98                   5.0                 14.45        3.19   
167                  4.0                 14.44        4.80   
53                   7.0                 14.35        2.84   
109                  3.0                 14.28        1.16   
150                  5.0                 14.25        1.28   
41                   3.0                 14.25        1.33   
95                   3.0                 14.22        1.63   
16                   4.0                 14.20        2.52   
94                   4.0                 13.96        3.09   
89                   3.0                 13.92        1.70   
174                  4.0                 13.88        1.94   
67                   4.0                 13.84        4.74   
175                  3.0                 13.44        4.35   
162                  5.0                 13.35        1.66   
83                   4.0                 13.32        1.02   
22                   6.0                 13.08        3.81   
119                  3.0                 12.96        4.43   
76                   4.0                 12.60        3.75   
172                  4.0                 12.56        4.24   
74                   3.0                 12.54        4.64   
117                  7.0                 12.53        1.79   
85                   7.0                 12.25        1.16   
79                   3.0                 12.15        2.96   
105                  4.0                 12.12        1.79   
182                  3.0                 12.09        4.40   
148                  3.0                 12.09        4.03   
44                   5.0                 11.90        2.73   
93                   6.0                 11.82        3.75   

                                        Item Name  Item ID  
92           Betrayal, Whisper of Grieving Widows       39  
178               Despair, Favor of Du

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



Purchase Count  \
Item ID Item Name                                                     
92      Final Critic                                             13   
178     Oathbreaker, Last Hope of the Breaking Storm             12   
82      Nirvana                                                   9   
145     Fiery Glass Crusader                                      9   
103     Singed Scalpel                                            8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

# Observable Trends
The total amount of players is 576

The vast majority of our users are male in the 20 to 24 age range. (214 out of 484)

Females are also mostly concentrated in the same age group. (40 out of 81)

Even though males (4.07) account for a much bigger total purchase value, females (4.47) and others (4.56) have a higher avg purchase per person 

The age range with highest purchase per person is 35 to 39 years old (4.76), surprisingly it is followed by than 10 years old ($ 4.54). 

64% of our total purchase value comes from ages 15 to 24. (1,526.95 of $ 2379.77)

Our highest individual total purchase value comes from user Lisosia93 with a total of $18.96


In [1024]:
# with pd.ExcelWriter('Resources/finalReport.xlsx') as writer:  

#     gender_dem_df.to_excel(
#         writer, sheet_name='Gender Demographics')

#     purchasing_analysis_g_df.to_excel(
#         writer, sheet_name='Purchasing Analysis (Gender)')
    
#     summary_demo_df.to_excel(
#         writer, sheet_name='Age Demographics')
    
#     summary_ages_df.to_excel(
#         writer, sheet_name='Purchasing Analysis (Age)')
    
#     top_spenders_df.to_excel(
#         writer, sheet_name='Top Spenders')